### PR TITLE
Clone web request's state

### PR DESCRIPTION
**Pull request by asvetlov**
Fix for #2284

### CODE REVIEW

File: aiohttp/web_request.py [hunk 1]
```diff
@@ -120,7 +123,8 @@ def clone(self, *, method=sentinel, rel_url=sentinel,
             self._time_service,
             self._task,
             self._loop,
-            secure_proxy_ssl_header=self._secure_proxy_ssl_header)
+            secure_proxy_ssl_header=self._secure_proxy_ssl_header,
+            state=self._state.copy())
```
Comment (kxepal) [inline aiohttp/web_request.py:127]: I hope there are no nested dicts/list in state, right?

**Comment by asvetlov** [comment]
It's impossible to perform deep state copy for an obvious reasons: values are not copyable very often, they could be an opened socket for example.

Thus we have 3 options:
1. Cleanup a state on copy.
2. Do swallow copy.
3. Share state between all copies.

I believe the option 2 is  better than others.

**Comment by kxepal** [comment]
@asvetlov What's a case for `req.clone()`? May be we can make it immutable instead to do not any copies?

**Comment by asvetlov** [comment]
`.clone()` is intended to be used in middlewares:
```
@asyncio.coroutine
def middleware_factory(app, next_handler):
    @asyncio.coroutine
    def middleware(request):
        request2 = request.clone(rel_url="/new_url", method="GET")
        response = yield from next_handler(request2)
    return middleware
```

I'd like to extend `clone()` params for fixing #2171 too.

**Comment by kxepal** [comment]
Hm...I see. Then it seems swallow copy is the only option.

**Pull request merged**